File: launch-video-templates/templates/19-monochrome-accent/qa.py
"""Extract finished-video scene and transition frames; verify delivery metadata."""

import json
import subprocess
from pathlib import Path

HERE = Path(__file__).resolve().parent
ROOT = HERE.parent.parent
OUT = ROOT / "out"
VIDEO = OUT / "19-monochrome-accent.mp4"
FRAMES = OUT / "19-monochrome-accent-frames"
SCENES = [
    ("HOOK", 60), ("CONTEXT", 192), ("BUILD + RUN", 340),
    ("SIMULATOR", 492), ("REPRODUCE / FIX / RETEST", 690),
    ("RECORDED EVIDENCE", 870), ("OUTCOME", 1050), ("LOGO", 1215),
]
BOUNDARIES = [135, 255, 420, 585, 765, 960, 1140]


def main():
    FRAMES.mkdir(exist_ok=True)
    probe = json.loads(subprocess.check_output([
        "ffprobe", "-v", "error", "-show_streams", "-show_format", "-of", "json", str(VIDEO),
    ]))
    video = next(stream for stream in probe["streams"] if stream["codec_type"] == "video")
    assert video["codec_name"] == "h264"
    assert (video["width"], video["height"]) == (1920, 1080)
    assert video["r_frame_rate"] == "30/1"
    assert video["pix_fmt"] == "yuv420p"
    assert int(video["nb_frames"]) == 1260
    assert abs(float(probe["format"]["duration"]) - 42) < 0.05
    (OUT / "19-monochrome-accent-ffprobe.json").write_text(json.dumps(probe, indent=2))
    samples = SCENES + [("FIRST FRAME", 0)]
    for boundary in BOUNDARIES:
        samples.extend([
            (f"BEFORE {boundary / 30:g}s", boundary - 1),
            (f"WIPE {boundary / 30:g}s", boundary + 6),
            (f"AFTER {boundary / 30:g}s", boundary + 33),
        ])
    samples.append(("FINAL FRAME", 1259))
    manifest = []
    for index, (label, frame) in enumerate(samples):
        target = FRAMES / f"{index:02d}.png"
        subprocess.run([
            "ffmpeg", "-v", "error", "-y", "-ss", f"{frame / 30:.8f}",
            "-i", str(VIDEO), "-frames:v", "1", str(target),
        ], check=True)
        manifest.append({"label": f"{label}  /  f{frame}", "path": str(target)})
    subprocess.run([
        "ffmpeg", "-v", "error", "-y", "-ss", "2", "-i", str(VIDEO),
        "-frames:v", "1", str(OUT / "19-monochrome-accent-poster.png"),
    ], check=True)
    manifest_path = OUT / "19-monochrome-accent-contact.json"
    manifest_path.write_text(json.dumps(manifest))
    subprocess.run([
        "swift", str(HERE / "contact_sheet.swift"), str(manifest_path),
        str(OUT / "19-monochrome-accent-contact-sheet.png"),
    ], check=True)
    print("PASS: H.264, 1920x1080, 30fps, yuv420p, 1260 frames, 42 seconds.")
    print(f"Extracted {len(samples)} full-resolution QA frames.")


if __name__ == "__main__":
    main()

File: launch-video-templates/templates/16-split-timeline/qa.mjs
import assert from 'node:assert/strict';
import {execFileSync} from 'node:child_process';
import {readFileSync} from 'node:fs';
import {dirname, resolve} from 'node:path';
import {fileURLToPath} from 'node:url';
import console from 'node:console';

const folder = dirname(fileURLToPath(import.meta.url));
const root = resolve(folder, '../..');
const manifest = JSON.parse(readFileSync(resolve(folder, 'template.json'), 'utf8'));
const video = resolve(root, 'out', `${manifest.id}.mp4`);
const probe = JSON.parse(execFileSync('ffprobe', [
  '-v', 'error', '-show_streams', '-show_format', '-of', 'json', video,
], {encoding: 'utf8'}));
const stream = probe.streams.find((entry) => entry.codec_type === 'video');
assert(stream, 'Missing video stream');
assert.equal(stream.codec_name, 'h264');
assert.equal(stream.width, manifest.width);
assert.equal(stream.height, manifest.height);
assert.equal(stream.r_frame_rate, `${manifest.fps}/1`);
assert.equal(Number(stream.nb_frames), manifest.fps * manifest.durationSeconds);
assert.equal(Number(stream.duration), manifest.durationSeconds);

const poster = resolve(root, 'out', `${manifest.id}-poster.png`);
execFileSync('ffmpeg', [
  '-y', '-hide_banner', '-loglevel', 'error', '-i', video,
  '-vf', 'select=eq(n\\,75)', '-frames:v', '1', poster,
]);

const sceneFrames = [75, 201, 330, 435, 495, 555, 651, 810, 1035, 1155];
const boundaries = [120, 240, 390, 570, 750, 930, 1080];
const frames = [...sceneFrames, ...boundaries.flatMap((frame) => [frame, frame + 12])].sort((a, b) => a - b);
const sheet = resolve(root, 'out', `${manifest.id}-contact-sheet.png`);
const selection = frames.map((frame) => `eq(n\\,${frame})`).join('+');
execFileSync('ffmpeg', [
  '-y', '-hide_banner', '-loglevel', 'error', '-i', video,
  '-vf', `select=${selection},scale=640:360,tile=3x8:padding=12:margin=12:color=0x102229`,
  '-frames:v', '1', sheet,
]);
console.log(JSON.stringify({
  codec: stream.codec_name,
  resolution: [stream.width, stream.height],
  fps: stream.r_frame_rate,
  frames: stream.nb_frames,
  videoDuration: stream.duration,
  containerDuration: probe.format.duration,
  poster,
  contactSheet: sheet,
  contactSheetFrameOrder: frames,
}, null, 2));
